---
title: "Replication of mQTLs associations across contexts"
author: Olivia Castellini-PÃ©rez
date: August 9, 2023
abstract: ""
output:  
  bookdown::pdf_book:
  keep_tex: true
  fig_caption: yes
  toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Models 

(1) n=all, data=bulk, model=interaction, cell-type=x4(salas), region=cis, cell-count=predicted
(2) n=all, data=bulk, model=interaction, cell-type=x4(salas), region=cis, cell-count=observed
(3) n=all, data=cell-specific, model=main, cell-type=x4(sorted), region=cis
(4) n=overlap(~60), data=cell-specific, model=main, cell-type=x4(sorted), region=cis
(5) n=overlap(~60), data=bulk, model=interaction, cell-type=x4(the same as sorted), region=cis

## Samples

|   Model   |Cell type |    N     |
|-----------|----------|----------|
| (1) & (2) |   All    |   1,086  |
|-----------|----------|----------|
|           |   Neu    |    158   |
|           |   Mono   |    141   |
|   (3)     |  Tcells  |    159   |
|           |  Bcells  |    100   |
|-----------|----------|----------|
|           |   Neu    |    75    |
|           |   Mono   |    60    |
| (4) & (5) |  Tcells  |    70    |
|           |  Bcells  |    45    |

## Comparisons 

```{r, include=FALSE}
library(dplyr)
library(data.table)
library(tidyr)
```

```{r, include=FALSE}
standardise <- function(d, ea="ea", oa="oa", beta="beta", chr="chr", pos="pos") {
    toflip <- d[[ea]] > d[[oa]]
    d[[beta]][toflip] <- d[[beta]][toflip] * -1
    temp <- d[[oa]][toflip]
    d[[oa]][toflip] <- d[[ea]][toflip]
    d[[ea]][toflip] <- temp
    d[["snpid"]] <- paste0(d[[chr]], ":", d[[pos]], "_", toupper(d[[ea]]), "_", toupper(d[[oa]]))
    d
}

repl_obs_exp <- function(b_disc, b_rep, se_disc, se_rep, alpha) {
  p_sign <- pnorm(-abs(b_disc) / se_disc) * pnorm(-abs(b_disc) / se_rep) + ((1 - pnorm(-abs(b_disc) / se_disc)) * (1 - pnorm(-abs(b_disc) / se_rep)))
  p_sig <- pnorm(-abs(b_disc) / se_rep + qnorm(alpha / 2)) + (1 - pnorm(-abs(b_disc) / se_rep - qnorm(alpha / 2)))
  p_rep <- pnorm(abs(b_rep)/se_rep, lower.tail=FALSE)
  res <- tibble::tibble(
    nsnp=length(b_disc),
    metric=c("Sign", "Sign", "P-value", "P-value"),
    datum=c("Expected", "Observed", "Expected", "Observed"),
    value=c(sum(p_sign, na.rm=TRUE), sum(sign(b_disc) == sign(b_rep)), sum(p_sig, na.rm=TRUE), sum(p_rep < alpha, na.rm=TRUE))
  )
  return(list(res=res, variants=dplyr::tibble(sig=p_sig, sign=p_sign)))
}

# Read in clumped
load("/home/uy23281/16_clumped.rdata")

# Organise data
clumped <- ungroup(clumped)
clumped$snpchr <- gsub("chr", "", clumped$snpchr)
clumped2 <- standardise(clumped, "Allele2", "Allele1", "Effect", "snpchr", "snppos")


##Loop to read and process all 100 chunks 
# Initialize an empty data frame to store the results
merged_data <- data.frame()

#Read standard deviation file 
sds<-read.table("/home/uy23281/model4/neu/EPIC_mQTL/input/sd.methy_neu.txt",header=T,sep="\t")

# Loop through 7 chunks of cell data
for (i in 1:7) {
  # Read in the current chunk
  chunk_file <- paste0("/home/uy23281/model4/neu/EPIC_mQTL/results/05/SortedNeu.res.cis.", i, ".RData")
  cell <- fread(chunk_file)
  cell <- left_join(cell, sds, by = c("gene" = "cpg"))

  # Subset the current chunk based on the clumped$cpg values
  cell <- subset(cell, gene %in% clumped$cpg)

  # Separate SNP column into chr and pos
  cell <- cell  %>% tidyr::separate(SNP, sep = ":", into = c("chr", "pos"))

  # Separate pos column into pos, a1, and a2
  cell <- cell %>% tidyr::separate(pos, sep = "_", into = c("pos", "a1", "a2"))

  # Standardize cell data
  cell2 <- standardise(cell, "a1", "a2", "beta", "chr", "pos")

  # Merge with clumped2 data
  m <- inner_join(clumped2, cell2, by = c("snpid", "cpg" = "gene"))

  # Calculate standard error
  m$se <- m$beta / m$`t-stat`

  # Append the current chunk's data to the merged_data
  if (i == 1) {
    merged_data <- m
  } else {
    merged_data <- rbind(merged_data, m)
  }
}
```

## Look at relationship between effects
```{r, echo=FALSE}
summary(lm(Effect ~ beta, merged_data))
```

## Look at overall replication rate
```{r, echo=FALSE}
table(merged_data$`p-value` < 0.05) %>% prop.table
```

## Look at replication rate (observed vs expected)
```{r, echo=FALSE}
repl<-repl_obs_exp(merged_data$Effect, merged_data$beta, merged_data$StdErr, merged_data$se, 5e-6)
write.table(repl,"/home/uy23281/model4/neu/EPIC_mQTL/results/repl_obs_exp.txt",quote=F)
repl
```
## Plot
```{r, echo=FALSE}
plot(Effect ~ beta, merged_data)
```




