---
title: "Replication of mQTLs associations across contexts"
author: Olivia Castellini-PÃ©rez
date: August 9, 2023
abstract: ""
output:  
  bookdown::pdf_book:
  keep_tex: true
  fig_caption: yes
  toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Models 

(1) n=all, data=bulk, model=interaction, cell-type=x4(salas), region=cis, cell-count=predicted
(2) n=all, data=bulk, model=interaction, cell-type=x4(salas), region=cis, cell-count=observed
(3) n=all, data=cell-specific, model=main, cell-type=x4(sorted), region=cis
(4) n=overlap(~60), data=cell-specific, model=main, cell-type=x4(sorted), region=cis
(5) n=overlap(~60), data=bulk, model=interaction, cell-type=x4(the same as sorted), region=cis

# Samples

|   Model   |Cell type |    N     |
|-----------|----------|----------|
| (1) & (2) |   All    |   1,086  |
|-----------|----------|----------|
|           |   Neu    |    158   |
|           |   Mono   |    141   |
|   (3)     |  Tcells  |    159   |
|           |  Bcells  |    100   |
|-----------|----------|----------|
|           |   Neu    |    75    |
|           |   Mono   |    60    |
| (4) & (5) |  Tcells  |    70    |
|           |  Bcells  |    45    |

# Replication

1. Are the GoDMC SNP-CpG pairs replicating in bulk+int?
2. Are the GoDMC SNP-CpG pairs replicating in cell-specific? 
3. Discovery in bulk+int replicating in cell-specific?  
4. Discovery in cell-specific replicating in bulk?

## GoDMC SNP-CpG pairs replication

```{r, include=FALSE}
library(dplyr)
library(data.table)
library(tidyr)
```

```{r, include=FALSE}
standardise <- function(d, ea="ea", oa="oa", beta="beta", chr="chr", pos="pos") {
    toflip <- d[[ea]] > d[[oa]]
    d[[beta]][toflip] <- d[[beta]][toflip] * -1
    temp <- d[[oa]][toflip]
    d[[oa]][toflip] <- d[[ea]][toflip]
    d[[ea]][toflip] <- temp
    d[["snpid"]] <- paste0(d[[chr]], ":", d[[pos]], "_", toupper(d[[ea]]), "_", toupper(d[[oa]]))
    d
}

repl_obs_exp <- function(b_disc, b_rep, se_disc, se_rep, alpha) {
  p_sign <- pnorm(-abs(b_disc) / se_disc) * pnorm(-abs(b_disc) / se_rep) + ((1 - pnorm(-abs(b_disc) / se_disc)) * (1 - pnorm(-abs(b_disc) / se_rep)))
  p_sig <- pnorm(-abs(b_disc) / se_rep + qnorm(alpha / 2)) + (1 - pnorm(-abs(b_disc) / se_rep - qnorm(alpha / 2)))
  p_rep <- pnorm(abs(b_rep)/se_rep, lower.tail=FALSE)
  res <- tibble::tibble(
    nsnp=length(b_disc),
    metric=c("Sign", "Sign", "P-value", "P-value"),
    datum=c("Expected", "Observed", "Expected", "Observed"),
    value=c(sum(p_sign, na.rm=TRUE), sum(sign(b_disc) == sign(b_rep)), sum(p_sig, na.rm=TRUE), sum(p_rep < alpha, na.rm=TRUE))
  )
  return(list(res=res, variants=dplyr::tibble(sig=p_sig, sign=p_sign)))
}

# Read in clumped
load("/user/home/uy23281/scratch/16_clumped.rdata")

# Organise data
clumped <- ungroup(clumped)
clumped$snpchr <- gsub("chr", "", clumped$snpchr)
clumped2 <- standardise(clumped, "Allele2", "Allele1", "Effect", "snpchr", "snppos")

```

### Model 4 
#### Relationship between GoDMC effect and cell interaction model 4 
```{r, include=FALSE}
##Loop to read and process all 100 chunks 
# Initialize an empty data frame to store the results

merged_data <- data.frame()
cell.list<-c("neu","mono","tcell","bcell")

#Create an empty list to store data frames and plots 
results_list <- list()
plots_list <- list()
prop_table_list <- list()
repl_list <- list()

#Loop through cell types
for (n in 1:length(cell.list)){
# Read standard deviation file
sds<-read.table(paste0("/user/home/uy23281/scratch/rmd.results/model4/",cell.list[n],"/sd.methy.txt"),header=T,sep="\t")
# Loop through 7 chunks of cell data
for (i in 1:7) {
  # Read in the current chunk
  chunk_file <- paste0("/user/home/uy23281/scratch/rmd.results/model4/",cell.list[n],"/res.cis.", i, ".gz")
  cell <- fread(chunk_file)
  cell <- left_join(cell, sds, by = c("gene" = "cpg"))

  # Subset the current chunk based on the clumped$cpg values
  cell <- subset(cell, gene %in% clumped$cpg)

  # Separate SNP column into chr and pos
  cell <- cell  %>% tidyr::separate(SNP, sep = ":", into = c("chr", "pos"))

  # Separate pos column into pos, a1, and a2
  cell <- cell %>% tidyr::separate(pos, sep = "_", into = c("pos", "a1", "a2"))

  # Standardize cell data
  cell2 <- standardise(cell, "a1", "a2", "beta", "chr", "pos")

  # Merge with clumped2 data
  m <- inner_join(clumped2, cell2, by = c("snpid", "cpg" = "gene"))

  # Calculate standard error
  m$se <- m$beta / m$`t-stat`

  # Append the current chunk's data to the merged_data
  if (i == 1) {
    merged_data <- m
  } else {
    merged_data <- rbind(merged_data, m)
  }
}
# 1 Relationship between the effects
full <- lm(Effect ~ beta, data = merged_data)
fitted <- summary(full)$coefficients
result_df <- data.frame(CellType = cell.list[n], Beta = fitted[2, 1], SE = fitted[2, 2],
                        PValue = fitted[2, 4])
#2 Overall replication rate
prop_table <- table(m$`p-value` < 0.05) %>% prop.table()
prop_table_df <- data.frame(CellType = cell.list[n], PropTable = prop_table)

#3 Replication rate 
repl <- repl_obs_exp(m$Effect, m$beta, m$StdErr, m$se, 5e-6)
repl_df <- data.frame(CellType = cell.list[n], Repl = repl)
  
#Save the different data frames 
results_list[[n]] <- result_df
prop_table_list[[n]] <- prop_table_df
repl_list[[n]] <- repl_df

# 4 Plots
plot <- ggplot(m, aes(x = beta, y = Effect)) +
    geom_point() +
    labs(title = paste("GoDMC effect vs model 4 effect for", cell.list[n]))
  
plots_list[[n]] <- plot
}

results_df <- do.call(rbind, results_list)
results_df<-as.data.table(results_df)
##  r function for formatting p-values

format.pvalue <- function(x) {
    formatted.x <- ifelse(x > 0.01,
                          formatC(x, digits = 2, format = "f"),
                          formatC(x, digits = 0))
    return(formatted.x)
}
results_df[, PValue := format.pvalue(PValue)]
results_df[, Beta := formatC(Beta, digits = 4, format = "f")]
results_df[, SE := formatC(SE, digits = 4, format = "f")]

#Save the plot 
grid_plots <- grid.arrange(grobs = plots_list, ncol = 2, nrow = 2)
ggsave(filename = "/user/home/uy23281/scratch/rmd.results/model4/Effect.plot.png", plot = grid_plots)
```

## Look at relationship between effects
```{r results_df, echo=FALSE}
knitr::kable(results_df, format="latex",digits=5,booktabs=T,
             col.names=c("Cell type","beta","SE","P-value"),align = "c",
             caption='\\textbf{Linear model results for the relationship between GoDMC and model 4 effects}')
```

## Look at overall replication rate
```{r prop_table_df, echo=FALSE}
print(prop_table_df)
```

## Look at replication rate (observed vs expected)
```{r repl_df, echo=FALSE}
print(repl_df)
```
## Plot
```{r plot, fig.align="center",fig.cap=c("GoDMC effect versus model 4 effect", echo=FALSE}
knitr::include_graphics("/user/home/uy23281/scratch/rmd.results/model4/Effect.plot.png")

```
